File: DSA/Stack/Code/PushOperation.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define MAX_SIZE 100

// Push Operation in C
typedef struct {
    int arr[MAX_SIZE];
    int top;
} Stack;

Stack* createStack() {
    Stack* stack = (Stack*)malloc(sizeof(Stack));
    stack->top = -1;
    return stack;
}

bool isFull(Stack* stack) {
    return stack->top == MAX_SIZE - 1;
}

void push(Stack* stack, int value) {
    if (isFull(stack)) {
        printf("Stack overflow - cannot push %d\n", value);
        return;
    }
    stack->arr[++stack->top] = value;
    printf("Successfully pushed: %d (top = %d)\n", value, stack->top);
}

void display(Stack* stack) {
    if (stack->top == -1) {
        printf("Stack is empty\n");
        return;
    }
    printf("Current stack: ");
    for (int i = 0; i <= stack->top; i++) {
        printf("%d ", stack->arr[i]);
    }
    printf("\n");
}

int main() {
    Stack* stack = createStack();
    
    printf("=== Push Operation ===\n\n");
    
    display(stack);
    
    for (int i = 1; i <= 5; i++) {
        push(stack, i * 10);
    }
    
    display(stack);
    
    printf("\nTrying to exceed capacity:\n");
    for (int i = 0; i < 96; i++) {
        push(stack, 999);
    }
    push(stack, 100); // This will fail
    
    printf("\nComplexity Analysis:\n");
    printf("Time: O(1)\n");
    printf("Space: O(1)\n");
    
    free(stack);
    return 0;
}

File: DSA/Stack/Code/PushOperation.py
# Push Operation in Python

class Stack:
    def __init__(self, max_size=100):
        self.arr = [0] * max_size
        self.top = -1
        self.max_size = max_size
    
    def is_full(self):
        return self.top == self.max_size - 1
    
    def push(self, value):
        if self.is_full():
            print(f"Stack overflow - cannot push {value}")
            return
        self.top += 1
        self.arr[self.top] = value
        print(f"Successfully pushed: {value} (top = {self.top})")
    
    def display(self):
        if self.top == -1:
            print("Stack is empty")
            return
        print("Current stack: ", end="")
        for i in range(self.top + 1):
            print(self.arr[i], end=" ")
        print()

# Test
stack = Stack()

print("=== Push Operation ===\n")

stack.display()

for i in range(1, 6):
    stack.push(i * 10)

stack.display()

print("\nComplexity Analysis:")
print("Time: O(1)")
print("Space: O(1)")
